sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,388 @@
+        Placeholder for the sums of the weights of every observation assigned
+        to each center.
+
+    labels : int array-like, shape (n_samples,)
+        labels assignment.
+
+    center_shift : {float32, float64} array-like, shape (n_clusters,)
+        Distance between old and new centers.
+
+    n_jobs : int
+        The number of threads to be used by openmp. If -1, openmp will use as
+        many as possible.
+
+    update_centers : bool
+        - If True, the labels and the new centers will be computed, i.e. runs
+          the E-step and the M-step of the algorithm.
+        - If False, only the labels will be computed, i.e runs the E-step of
+          the algorithm.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int n_clusters = centers_new.shape[0]
+
+        # hard-coded number of samples per chunk. Appeared to be close to
+        # optimal in all situations.
+        int n_samples_chunk = 256 if n_samples > 256 else n_samples
+        int n_chunks = n_samples // n_samples_chunk
+        int n_samples_r = n_samples % n_samples_chunk
+        int chunk_idx, n_samples_chunk_eff
+        int start, end
+        # int num_threads
+
+        int j, k
+
+    # If n_samples < 256 there's still one chunk of size n_samples_r
+    if n_chunks == 0:
+        n_chunks = 1
+        n_samples_chunk = 0
+
+    # re-initialize all arrays at each iteration
+    centers_squared_norms = row_norms(centers_new, squared=True)
+
+    if update_centers:
+        memcpy(&centers_old[0, 0], &centers_new[0, 0], n_clusters * n_features * sizeof(floating))
+        memset(&centers_new[0, 0], 0, n_clusters * n_features * sizeof(floating))
+        memset(&weight_in_clusters[0], 0, n_clusters * sizeof(floating))
+
+    # set number of threads to be used by openmp
+    # num_threads = n_jobs if n_jobs != -1 else openmp.omp_get_max_threads()
+
+    for chunk_idx in prange(n_chunks, nogil=True, num_threads=n_jobs):
+        # remaining samples added to last chunk
+        if chunk_idx == n_chunks - 1:
+            n_samples_chunk_eff = n_samples_chunk + n_samples_r
+        else:
+            n_samples_chunk_eff = n_samples_chunk
+
+        start = chunk_idx * n_samples_chunk
+        end = start + n_samples_chunk_eff
+
+        _update_chunk_dense(
+            &X[start, 0],
+            sample_weight[start: end],
+            x_squared_norms[start: end],
+            centers_old,
+            centers_new,
+            centers_squared_norms,
+            weight_in_clusters,
+            labels[start: end],
+            update_centers)
+
+    if update_centers:
+        _relocate_empty_clusters_dense(X, sample_weight, centers_old,
+                                       centers_new, weight_in_clusters, labels)
+
+        _average_centers(centers_new, weight_in_clusters)
+        _center_shift(centers_old, centers_new, center_shift)
+
+
+cdef void _update_chunk_dense(floating *X,
```

Comment:
Maybe add short comment about X being a pointer on a C-aligned 2D array

Reply:
I think X is enough explicit. Even if here it's a pointer, X means the same thing throughout the whole code base.

Reply:
@jeremiedbb replied in the main thread the following. Let me consolidate it here instead:

> That would not hurt to be explicit. The fact that it's expected to be C-aligned in particular is not trivial.

What do you mean ? Adding a comment ?

> BTW wouldn't be possible to use a typed memory view here to enforce this and then retrieve the pointer in the body of the function?

Not yet because of the issue between fused typed memoryview and read only arrays. It's fixed on cython master and might be available for the next cython release. I don't know if we are ready to bump the min cython version to the last one.

Reply:
Yes adding a comment to explain that this is expected to be a point to a C-aligned 2D array.

Ok for not using a fused typed memoryview as long as the oldest Cython versions supported has this issue. Please add a comment to state that we should update that at some point in the future.